File: protocols/proven-logcollector/ffi/zig/src/logcollector.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// logcollector.zig — Zig FFI implementation of proven-logcollector.
//
// Implements the structured log collection pipeline primitive with:
//   - Slot-based pipeline management (up to 64 concurrent pipelines)
//   - Pipeline stage state machine (Input -> Parse -> Filter -> Transform -> Output)
//   - Log level filtering with configurable minimum threshold
//   - Entry ingestion with level-based filtering
//   - Filter operation tracking
//   - Thread-safe via mutex on global state
//
// Tag values MUST match:
//   - Idris2 ABI (src/LogcollectorABI/Layout.idr)
//   - C header   (generated/abi/logcollector.h)

const std = @import("std");

// ── Enums (matching Idris2 Layout.idr tag assignments exactly) ──────────

/// LogLevel — matches logLevelToTag
pub const LogLevel = enum(u8) {
    trace = 0,
    debug = 1,
    info = 2,
    warn = 3,
    err = 4,
    fatal = 5,
};

/// InputFormat — matches inputFormatToTag
pub const InputFormat = enum(u8) {
    json = 0,
    logfmt = 1,
    syslog = 2,
    cef = 3,
    gelf = 4,
    raw = 5,
};

/// OutputTarget — matches outputTargetToTag
pub const OutputTarget = enum(u8) {
    file = 0,
    elasticsearch = 1,
    s3 = 2,
    kafka = 3,
    stdout = 4,
};

/// FilterOp — matches filterOpToTag
pub const FilterOp = enum(u8) {
    include = 0,
    exclude = 1,
    transform = 2,
    redact = 3,
    sample = 4,
};

/// PipelineStage — matches pipelineStageToTag
pub const PipelineStage = enum(u8) {
    input = 0,
    parse = 1,
    filter = 2,
    pipeline_transform = 3,
    output = 4,
};

/// LogcollectorError — matches logcollectorErrorToTag
pub const LogcollectorError = enum(u8) {
    ok = 0,
    invalid_slot = 1,
    not_active = 2,
    invalid_transition = 3,
    below_threshold = 4,
    capacity_exhausted = 5,
    invalid_param = 6,
};

// ── Pipeline Context instance ───────────────────────────────────────────

const PipelineCtx = struct {
    /// Whether this slot is in use.
    active: bool,
    /// Input format for this pipeline.
    input_format: InputFormat,
    /// Output destination.
    output_target: OutputTarget,
    /// Minimum log level to accept (entries below are dropped).
    min_level: LogLevel,
    /// Current pipeline stage.
    current_stage: PipelineStage,
    /// Last error code (255 = no error).
    last_error: u8,
    /// Number of log entries successfully processed.
    entries_processed: u32,
    /// Number of log entries dropped (below threshold or filtered).
    entries_dropped: u32,
    /// Number of filter operations applied.
    filters_applied: u32,
};

// ── Global state (slot-based, mutex-protected) ──────────────────────────

const MAX_CONTEXTS: usize = 64;

const empty_ctx: PipelineCtx = .{
    .active = false,
    .input_format = .json,
    .output_target = .stdout,
    .min_level = .trace,
    .current_stage = .input,
    .last_error = 255,
    .entries_processed = 0,
    .entries_dropped = 0,
    .filters_applied = 0,
};

var contexts: [MAX_CONTEXTS]PipelineCtx = [_]PipelineCtx{empty_ctx} ** MAX_CONTEXTS;
var mutex: std.Thread.Mutex = .{};

// ── Helpers ─────────────────────────────────────────────────────────────

fn getActive(slot: c_int) ?*PipelineCtx {
    if (slot < 0 or slot >= MAX_CONTEXTS) return null;
    const idx: usize = @intCast(slot);
    if (!contexts[idx].active) return null;
    return &contexts[idx];
}

// ── ABI version ─────────────────────────────────────────────────────────

pub export fn lc_abi_version() callconv(.c) u32 {
    return 1;
}

// ── Lifecycle ───────────────────────────────────────────────────────────

/// Create a new log collection pipeline.
pub export fn lc_create(input_fmt: u8, output_target: u8, min_level: u8) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (input_fmt > 5) return -1;
    if (output_target > 4) return -1;
    if (min_level > 5) return -1;

    for (&contexts, 0..) |*ctx, i| {
        if (!ctx.active) {
            ctx.* = empty_ctx;
            ctx.active = true;
            ctx.input_format = @enumFromInt(input_fmt);
            ctx.output_target = @enumFromInt(output_target);
            ctx.min_level = @enumFromInt(min_level);
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a log collection pipeline.
pub export fn lc_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();

    if (slot < 0 or slot >= MAX_CONTEXTS) return;
    const idx: usize = @intCast(slot);
    contexts[idx].active = false;
}

// ── State queries ───────────────────────────────────────────────────────

pub export fn lc_get_input_format(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.input_format);
}

pub export fn lc_get_output_target(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.output_target);
}

pub export fn lc_get_min_level(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.min_level);
}

pub export fn lc_get_current_stage(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.current_stage);
}

pub export fn lc_get_entries_processed(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const ctx = getActive(slot) orelse return 0;
    return ctx.entries_processed;
}

pub export fn lc_get_entries_dropped(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const ctx = getActive(slot) orelse return 0;
    return ctx.entries_dropped;
}

pub export fn lc_get_last_error(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const ctx = getActive(slot) orelse return 255;
    return ctx.last_error;
}

// ── Log ingestion ───────────────────────────────────────────────────────

/// Ingest a log entry with the given level.
/// Returns LogcollectorError tag.
/// Entries below min_level are dropped (BelowThreshold).
pub export fn lc_ingest(slot: c_int, level: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(LogcollectorError.invalid_slot);

    if (level > 5) {
        ctx.last_error = @intFromEnum(LogcollectorError.invalid_param);
        return @intFromEnum(LogcollectorError.invalid_param);
    }

    // Check against minimum level threshold
    if (level < @intFromEnum(ctx.min_level)) {
        ctx.entries_dropped += 1;
        ctx.last_error = @intFromEnum(LogcollectorError.below_threshold);
        return @intFromEnum(LogcollectorError.below_threshold);
    }

    ctx.entries_processed += 1;
    ctx.last_error = 255;
    return @intFromEnum(LogcollectorError.ok);
}

// ── Filter operations ───────────────────────────────────────────────────

/// Apply a filter operation to the pipeline.
/// Returns LogcollectorError tag.
pub export fn lc_apply_filter(slot: c_int, filter_op: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(LogcollectorError.invalid_slot);

    if (filter_op > 4) {
        ctx.last_error = @intFromEnum(LogcollectorError.invalid_param);
        return @intFromEnum(LogcollectorError.invalid_param);
    }

    ctx.filters_applied += 1;
    ctx.last_error = 255;
    return @intFromEnum(LogcollectorError.ok);
}

// ── Pipeline stage advancement ──────────────────────────────────────────

/// Advance the pipeline to the next stage.
/// Stages must advance in order: Input -> Parse -> Filter -> Transform -> Output.
/// Returns LogcollectorError tag.
pub export fn lc_advance_stage(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(LogcollectorError.invalid_slot);

    const current = @intFromEnum(ctx.current_stage);
    if (current >= 4) {
        // Already at Output, cannot advance further
        ctx.last_error = @intFromEnum(LogcollectorError.invalid_transition);
        return @intFromEnum(LogcollectorError.invalid_transition);
    }

    ctx.current_stage = @enumFromInt(current + 1);
    ctx.last_error = 255;
    return @intFromEnum(LogcollectorError.ok);
}

// ── Configuration setters ───────────────────────────────────────────────

/// Set the minimum log level for filtering.
pub export fn lc_set_min_level(slot: c_int, level: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(LogcollectorError.invalid_slot);

    if (level > 5) {
        ctx.last_error = @intFromEnum(LogcollectorError.invalid_param);
        return @intFromEnum(LogcollectorError.invalid_param);
    }

    ctx.min_level = @enumFromInt(level);
    ctx.last_error = 255;
    return @intFromEnum(LogcollectorError.ok);
}

File: bindings/rust/src/logcollector.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//

//! Log Collector types for the proven-servers ABI.
//!
//! Formally verified log collection/pipeline types.
//! Mirrors the Idris2 module `LogcollectorABI.Types`.
//!
//! - `LogLevel` -- Log severity levels.
//! - `InputFormat` -- Log input formats.
//! - `OutputTarget` -- Log output targets.
//! - `FilterOp` -- Log filter operations.
//! - `PipelineStage` -- Log pipeline stages.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// LogLevel (tags 0-5)
// ===========================================================================

/// Log severity levels.
///
/// Matches `LogLevel` in `LogcollectorABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum LogLevel {
    /// Trace (tag 0).
    Trace = 0,
    /// Debug (tag 1).
    Debug = 1,
    /// Info (tag 2).
    Info = 2,
    /// Warn (tag 3).
    Warn = 3,
    /// Error (tag 4).
    Err = 4,
    /// Fatal (tag 5).
    Fatal = 5,
}

impl LogLevel {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Trace),
            1 => Some(Self::Debug),
            2 => Some(Self::Info),
            3 => Some(Self::Warn),
            4 => Some(Self::Err),
            5 => Some(Self::Fatal),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [LogLevel; 6] = [
        Self::Trace, Self::Debug, Self::Info, Self::Warn, Self::Err, Self::Fatal,
    ];
}

impl fmt::Display for LogLevel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// InputFormat (tags 0-5)
// ===========================================================================

/// Log input formats.
///
/// Matches `InputFormat` in `LogcollectorABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum InputFormat {
    /// JSON (tag 0).
    Json = 0,
    /// Logfmt (tag 1).
    Logfmt = 1,
    /// Syslog (tag 2).
    Syslog = 2,
    /// CEF (tag 3).
    Cef = 3,
    /// GELF (tag 4).
    Gelf = 4,
    /// Raw (tag 5).
    Raw = 5,
}

impl InputFormat {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Json),
            1 => Some(Self::Logfmt),
            2 => Some(Self::Syslog),
            3 => Some(Self::Cef),
            4 => Some(Self::Gelf),
            5 => Some(Self::Raw),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [InputFormat; 6] = [
        Self::Json, Self::Logfmt, Self::Syslog, Self::Cef, Self::Gelf, Self::Raw,
    ];
}

impl fmt::Display for InputFormat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// OutputTarget (tags 0-4)
// ===========================================================================

/// Log output targets.
///
/// Matches `OutputTarget` in `LogcollectorABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum OutputTarget {
    /// File (tag 0).
    File = 0,
    /// Elasticsearch (tag 1).
    Elasticsearch = 1,
    /// S3 (tag 2).
    S3 = 2,
    /// Kafka (tag 3).
    Kafka = 3,
    /// Stdout (tag 4).
    Stdout = 4,
}

impl OutputTarget {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::File),
            1 => Some(Self::Elasticsearch),
            2 => Some(Self::S3),
            3 => Some(Self::Kafka),
            4 => Some(Self::Stdout),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [OutputTarget; 5] = [
        Self::File, Self::Elasticsearch, Self::S3, Self::Kafka, Self::Stdout,
    ];
}

impl fmt::Display for OutputTarget {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// FilterOp (tags 0-4)
// ===========================================================================

/// Log filter operations.
///
/// Matches `FilterOp` in `LogcollectorABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum FilterOp {
    /// Include (tag 0).
    Include = 0,
    /// Exclude (tag 1).
    Exclude = 1,
    /// Transform (tag 2).
    Transform = 2,
    /// Redact (tag 3).
    Redact = 3,
    /// Sample (tag 4).
    Sample = 4,
}

impl FilterOp {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Include),
            1 => Some(Self::Exclude),
            2 => Some(Self::Transform),
            3 => Some(Self::Redact),
            4 => Some(Self::Sample),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [FilterOp; 5] = [
        Self::Include, Self::Exclude, Self::Transform, Self::Redact, Self::Sample,
    ];
}

impl fmt::Display for FilterOp {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// PipelineStage (tags 0-4)
// ===========================================================================

/// Log pipeline stages.
///
/// Matches `PipelineStage` in `LogcollectorABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum PipelineStage {
    /// Input (tag 0).
    Input = 0,
    /// Parse (tag 1).
    Parse = 1,
    /// Filter (tag 2).
    Filter = 2,
    /// Transform (tag 3).
    PipelineTransform = 3,
    /// Output (tag 4).
    Output = 4,
}

impl PipelineStage {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Input),
            1 => Some(Self::Parse),
            2 => Some(Self::Filter),
            3 => Some(Self::PipelineTransform),
            4 => Some(Self::Output),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [PipelineStage; 5] = [
        Self::Input, Self::Parse, Self::Filter, Self::PipelineTransform, Self::Output,
    ];
}

impl fmt::Display for PipelineStage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn log_level_roundtrip() {
        for v in LogLevel::ALL {
            let tag = v.to_tag();
            let decoded = LogLevel::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(LogLevel::from_tag(6).is_none());
    }

    #[test]
    fn input_format_roundtrip() {
        for v in InputFormat::ALL {
            let tag = v.to_tag();
            let decoded = InputFormat::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(InputFormat::from_tag(6).is_none());
    }

    #[test]
    fn output_target_roundtrip() {
        for v in OutputTarget::ALL {
            let tag = v.to_tag();
            let decoded = OutputTarget::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(OutputTarget::from_tag(5).is_none());
    }

    #[test]
    fn filter_op_roundtrip() {
        for v in FilterOp::ALL {
            let tag = v.to_tag();
            let decoded = FilterOp::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(FilterOp::from_tag(5).is_none());
    }

    #[test]
    fn pipeline_stage_roundtrip() {
        for v in PipelineStage::ALL {
            let tag = v.to_tag();
            let decoded = PipelineStage::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(PipelineStage::from_tag(5).is_none());
    }

}
